packages/core/contracts/oracle/implementation/EmergencyProposer.sol

```diff
@@ -0,0 +1,238 @@
+// SPDX-License-Identifier: AGPL-3.0-only
+pragma solidity 0.8.16;
+
+import "./Finder.sol";
+import "./GovernorV2.sol";
+import "./Constants.sol";
+import "../interfaces/OracleAncillaryInterface.sol";
+import "./AdminIdentifierLib.sol";
+import "../../common/implementation/Lockable.sol";
+import "@openzeppelin/contracts/access/Ownable.sol";
+import "@openzeppelin/contracts/token/ERC20/IERC20.sol";
+import "@openzeppelin/contracts/token/ERC20/utils/SafeERC20.sol";
+
+/**
+ * @title Emergency Proposer contract
+ * @dev This is a contract that allows anyone to construct an emergency recovery transaction to bypass the
+ * standard voting process by submitting a very large bond, which is considered a quorum in this case. This bond is
+ * expected to be about as large as the GAT in the VotingV2 contract. If a proposal is considered invalid, UMA token
+ * holders can vote to slash and remove this proposal through the standard governance flow. If valid, a proposal must
+ * wait minimumWaitTime before it can be executed and it can only be executed by a privileged account, executor. This
+ * includes three tiers of protection to ensure that abuse is extremely risky both from creating market volatility in
+ * the underlying token and the threat of the locked tokens being slashed.
+ */
+contract EmergencyProposer is Ownable, Lockable {
+    using SafeERC20 for IERC20;
+    IERC20 public immutable token;
+    uint256 public quorum;
+    uint64 public minimumWaitTime;
+
+    GovernorV2 public immutable governor;
+    Finder public immutable finder;
+
+    struct EmergencyProposal {
+        address sender;
+        uint64 expiryTime;
+        uint256 lockedTokens;
+        GovernorV2.Transaction[] transactions;
+    }
+    mapping(uint256 => EmergencyProposal) public emergencyProposals;
```

Comment:
reason to use a mapping like this over an array?

Reply:
Nope. I just started with the ProposerV2 as a template, so that was inherited. Agree an array makes more sense here. Will change.

Reply:
Done.